HPCC-27051 create a sasha service to clean up post mortem files

streeterd opened the pull request:
<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [x] Send notifications about my Pull Request position in Smoketest queue.
- [x] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

streeterd commented on dali/sasha/sadbghk.cpp:
```diff
@@ -0,0 +1,228 @@
+        }
+        else
+        {
+            PROGLOG(LOGDBGHK "Non post-mortem dir: %s", dirName.str());
```
@jakesmith do we need this log?
I added it in case there were folders that we not expected.

jakesmith replied:
well isPostMortemDirPath is needed so moot point.

streeterd commented on dali/sasha/sadbghk.cpp:
```diff
@@ -0,0 +1,228 @@
+        RegExpr postMortemDirRegEx("^W[0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9]-[0-9][0-9][0-9][0-9][0-9][0-9]$");
+        if (postMortemDirRegEx.find(dirName.str()))
+        {
+            PROGLOG(LOGDBGHK "Post-mortem dir: %s", dirName.str());
```
@jakesmith do we need this log?
Could it be useful as a DBGLOG ?

jakesmith replied:
we need to be very careful not to be too verbose - logging is expensive.
DBGLOG would be 'ok' in theory, but in fact there our default logging level in various places is still too high and would include DBGLOG's (DBGLOG confusingly is emitted in release too).

You do truly want it to be a DEBUG only log, during development only. For now, we should either wrap them in #ifdef _DEBUG if you really want to leave them in code, or remove them entirely when done.

streeterd commented on dali/sasha/sadbghk.cpp:
```diff
@@ -0,0 +1,228 @@
+
+        if (now.compare(expires, false) > 0)
+        {
+            PROGLOG(LOGDBGHK "Post-mortem dir: %s has expired", dirName.str());
```
@jakesmith is this log needed?
The dir will be deleted and logged anyway.

jakesmith replied:
right, ensure logging by default is at a minimum, logging when deleted is more than enough. So this func can just return true/false only.

streeterd commented on helm/hpcc/values.yaml:
```diff
@@ -513,6 +513,13 @@ sasha:
     #switchMinTime: 1
     #queues: "*"
 
+  debug-housekeeping: {} # NB: no properties defined, use defaults
```
The new code matches the existing code.

jakesmith commented on dali/sasha/sadbghk.cpp:
```diff
@@ -0,0 +1,224 @@
+        }
+    }
+
+    bool isExpiredDirPath(StringBuffer &dirName, const unsigned &defaultExpireDays)
```
see other comments, should use iFile and iFile.getTime to look at modified-time.

jakesmith commented on dali/sasha/sadbghk.cpp:
```diff
@@ -0,0 +1,224 @@
+{
+    assertex(!sashaDebugHousekeepingServer); // initialization problem
+    Linked<IPropertyTree> config = serverConfig;
+    sashaDebugHousekeepingServer = new CSashaDebugHousekeepingServer(config);
```
see other comments re. containerized, and using getComponentConfig*(). SO no need to pass in config here.
Worth adding a assertex(isContainerized() in this func for clarity.

jakesmith commented on dali/sasha/saserver.cpp:
```diff
@@ -435,6 +437,8 @@ int main(int argc, const char* argv[])
                     servers.append(*createSashaFileExpiryServer());
                 else if (strieq(service, "thor-qmon"))
                     servers.append(*createSashaQMonitorServer());
+                else if (strieq(service, "debug-housekeeping"))
```
let's rename to "debugplane-housekeeping"

jakesmith commented on dali/sasha/saserver.cpp:
```diff
@@ -75,6 +76,7 @@ static void AddServers()
     servers.append(*createSashaDaFSMonitorServer());
     servers.append(*createSashaQMonitorServer());
     servers.append(*createSashaFileExpiryServer()); 
+    servers.append(*createSashaDebugHousekeepingServer());
```
this is BM only, and this service will not be useable there, this line should be removed.

jakesmith commented on docs/EN_US/ContainerizedHPCC/ContainerizedMods/LocalDeployment.xml:
```diff
@@ -99,6 +99,7 @@ dali.sasha.coalescer
 sasha.dfurecovery-archiver
 sasha.dfuwu-archiver
 sasha.file-expiry
+sasha.debug-housekeeping
```
debugplane-housekeeping .. and in other yaml's

jakesmith commented on docs/EN_US/ConfiguringHPCC/ConfiguringHPCC.xml:
```diff
@@ -2417,6 +2417,29 @@ sudo -u hpcc cp /etc/HPCCSystems/source/NewEnvironment.xml /etc/HPCCSystems/envi
 
         <?hard-pagebreak ?>
 
+        <sect3>
```
docs/*/ConfiguringHPCC/ are BM only docs, should be left unchanged as this service will not be part of BM.

jakesmith commented on helm/hpcc/templates/_helpers.tpl:
```diff
@@ -1715,6 +1715,8 @@ dali
 {{- else if (eq "thor-qmon" .name) -}}
 dali queues
+{{- else if (eq "debug-housekeeping" .name) -}}
+dali data
```
does not need 'dali' access.

jakesmith commented on helm/hpcc/values.schema.json:
```diff
@@ -3045,6 +3045,35 @@
+        },
+        "user": {
+          "type": "string",
+          "description": "A username authorized to access and remove post-mortem files"
```
does not need a 'user'.

jakesmith commented on helm/hpcc/values.schema.json:
```diff
@@ -3045,6 +3045,35 @@
+      "type": "object",
+      "allOf": [{ "$ref": "#/definitions/sashacommon" }],
+      "properties": {
+        "expiryDefault": {
```
see other comment, let's rename to "expiryDays'

jakesmith commented on testing/helm/tests/resourced.yaml:
```diff
@@ -510,6 +510,16 @@ sasha:
+    #user: sasha
+    resources:
+      cpu: "1"
+      memory: "4Gi"
```
this service in particular should require a trivial amount of memory, to emphasis that, let's reduce this to 1Gi

jakesmith reviewed (changes requested):
@streeterd - please see comments

jakesmith commented on helm/hpcc/values.schema.json:
```diff
@@ -3045,6 +3045,31 @@
+      "properties": {
+        "expiryDays": {
+          "type": "integer",
+          "description": "Default number of days passed that post-mortem files are deleted after their associated workunit was completed",
```
'after their associated workunit was completed' doesn't really make sense.

jakesmith commented on helm/hpcc/values.schema.json:
```diff
@@ -3045,6 +3045,31 @@
+        },
+        "disabled": {},
+        "interval": {},
+        "service": {},
```
I can't remember how these work exactly, I think they're necessary if you want to inherit them in the schema,
but "service" we don't want as part of this sasha component.
Others that are not relevant:
- throttle
- plane
- egress

jakesmith commented on helm/hpcc/values.yaml:
```diff
@@ -513,6 +513,12 @@ sasha:
 
+  debugplane-housekeeping: {} # NB: no properties defined, use defaults
+    #disabled: true
+    #interval: 1
```
let's make this 24 - in keeping with the default in the code (and in other example/test yamls)

jakesmith commented on dali/sasha/sadbghk.cpp:
```diff
@@ -0,0 +1,164 @@
+#include "jstring.hpp"
+
+#include "dadfs.hpp"
+#include "dalienv.hpp"
```
these ^^ two aren't needed I think.

streeterd replied:
dadfs.hpp is needed for IFile and IDirectoryIterator
dalienv.hpp is needed for the LOGs

jakesmith replied:
that's not where those things are defined.

jakesmith commented on dali/sasha/sadbghk.cpp:
```diff
@@ -0,0 +1,164 @@
+{
+    bool stopped{false};
+    Semaphore stopsem;
+    Mutex runmutex;
```
super picky: camelCase prefered (for stopSem and runMutex)

jakesmith commented on dali/sasha/sadbghk.cpp:
```diff
@@ -0,0 +1,164 @@
+
+class CSashaDebugPlaneHousekeepingServer : public ISashaServer, public Thread
+{
+    bool stopped{false};
```
should really be an atomic

jakesmith commented on dali/sasha/sadbghk.cpp:
```diff
@@ -0,0 +1,164 @@
+            return false;
+    }
+
+} *sashaDebugPlaneHousekeepingServer = NULL;
```
picky!: nullptr preferred

jakesmith commented on dali/sasha/sadbghk.hpp:
```diff
@@ -0,0 +1,7 @@
+#ifndef SADBGHKIF_HPP
+#define SADBGHKIF_HPP
```
trivial: I know lots of code uses this approach, but we should use this instead:
```
#pragma once
```

jakesmith reviewed (commented):
@streeterd - looking close - please see comments.

jakesmith commented on dali/sasha/sadbghk.cpp:
```diff
@@ -83,41 +83,42 @@ class CSashaDebugPlaneHousekeepingServer : public ISashaServer, public Thread
+                // Process directories, exclude the ".", "..", and non expired directories from housekeeping
                 if (streq(filePath, ".") || streq(filePath, ".."))
-                    continue;
+                        continue;
```
extra indentation.

jakesmith commented on dali/sasha/sadbghk.cpp:
```diff
@@ -0,0 +1,168 @@
+    Semaphore stopSem;
+    Mutex runMutex;
+    StringBuffer debugDir;
+    unsigned expiryDays{0};
```
let's add a static constexpr unsigned defaultExpiryDays = 14; and use here and on line 40 too.
And in start (before Thread::start) check that if set to 0, then bails out (doesn't start).

In fact the "Failed to get default debug plane" bail out should before the Thread::start(false) too.
Move the Thread::start(false) to the end of the function.

jakesmith commented on dali/sasha/sadbghk.cpp:
```diff
@@ -0,0 +1,168 @@
+                        continue;
+
+                if (hasExpired(iFile))
+                {
```
style/personal: don't normally add block curlies for single line blocks.

jakesmith commented on helm/hpcc/values.yaml:
```diff
@@ -513,6 +513,12 @@ sasha:
+    #disabled: true
+    #interval: 24
+    #at: "* 3 * * *"
+    #expiryDays: 7
```
for consistency, make this 14, i.e. so same as stated default in schema.

jakesmith commented on testing/helm/tests/resourced.yaml:
```diff
@@ -510,6 +510,15 @@ sasha:
+    #disabled: true
+    #interval: 1
+    #at: "* 3 * * *"
+    #expiryDays: 7
```
for consistency, make this 14, i.e. so same as stated default in schema.

jakesmith commented on testing/helm/tests/resourced2.yaml:
```diff
@@ -512,6 +512,15 @@ sasha:
+    #disabled: true
+    #interval: 1
+    #at: "* 3 * * *"
+    #expiryDays: 7
```
for consistency, make this 14, i.e. so same as stated default in schema.

jakesmith reviewed (commented):
@streeterd - please see comments.

streeterd commented on dali/sasha/sadbghk.cpp:
```diff
@@ -137,11 +137,15 @@ class CSashaDebugPlaneHousekeepingServer : public ISashaServer, public Thread
         CDateTime expiredModifiedTime;
-        file->getTime(nullptr, &expiredModifiedTime, nullptr);
+        iFile.getTime(nullptr, &expiredModifiedTime, nullptr);
+        if (expiredModifiedTime.isNull())
```
Not that I can remember, I will use the return of `getTime`

streeterd commented:
> @streeterd - please see comments.

Fixed in a4adb7505207c5a52be8cc1c6cace2ff1158bff7

streeterd committed:
HPCC-27051 Add Sasha service to clean debug plane

Create a new Sasha housekeeping service to delete post-mortem files and
directories on the debug plane.

Signed-off-by: Dave Streeter <dave.streeter@lexisnexisrisk.com>

jakesmith reviewed (commented):
@streeterd - looks good. Please squash.

After that, happy for it to be merged and then add a separate linked documentation JIRA, or could add the documentation md in this PR if you prefer.

jakesmith reviewed (commented):
@streeterd - looks good, but you're previous commit with a table detailing the config options looked clearer I think.

jakesmith commented on dali/sasha/sasha.md:
```diff
@@ -0,0 +1,66 @@
+
+1. The `disabled` value is read from the component configuration
+2. If `disabled` is `true` then the service is not started.
+3. If the service is being run in the bare metal environment then the service is stopped.
```
Should remove this line. The service is never created in BM

jakesmith commented on dali/sasha/sasha.md:
```diff
@@ -0,0 +1,66 @@
+5. The debug plane location is read from the component configuration.
+6. The service `Thread` is started.
+7. The schedule `at` is read from the component configuration.
+8. The `interval` is read from the component configuration.
```
Step 4, 7 and 8 are a bit OTT to say config option is read from config file..

jakesmith commented on dali/sasha/sadbghk.cpp:
```diff
@@ -0,0 +1,180 @@
+        expiryDays = getComponentConfigSP()->getPropInt("@expiryDays", defaultExpiryDays);
+        if (expiryDays == 0)
+        {
+            WARNLOG("Failed to get expiry days for debug plane");
```
Not "failed to get" - it had to be deliberately configured as 0 for this to be true.
I would treat as disabled and exit without starting.

In fact a better pattern is to set 'stopped' to true in member initialization,
and set to false just before the Thread::start() call.

It would also be good to check 'interval' here for 0, and not start, rather than check it once started and then stop.

jakesmith commented on dali/sasha/sadbghk.cpp:
```diff
@@ -0,0 +1,180 @@
+
+    virtual void stop() override
+    {
+        if (!stopped)
```
I know this is c&p from other code, but if already stopped, why continue..
I think this should be:
```
if (stopped)
    return;
stopped = true;
stopSem.signal();
....
```

jakesmith reviewed (commented):
@streeterd - spotted a couple of minors worth addressing before this is merged.

jakesmith commented on dali/sasha/sadbghk.cpp:
```diff
@@ -35,12 +36,9 @@ class CSashaDebugPlaneHousekeepingServer : public ISashaServer, public Thread
-        }
+        Owned<IPropertyTree> compConfig = getComponentConfig();
+        expiryDays = compConfig->getPropInt("@expiryDays", defaultExpiryDays);
+        interval = compConfig->getPropInt("@interval", defaultHouseKeepingIntervalHours);
```
minor: cleaner to check these here and exit (return), and then unconditionally set stopped to false, just before thread starts.

jakesmith reviewed (commented):
@streeterd - looks good, 1 trivial comment.

jakesmith reviewed (commented):
@streeterd - looks good. Please squash.

streeterd commented:
@jakesmith  squashed.

jakesmith reviewed (dismissed):
@streeterd - looks good to go.

jakesmith commented:
@streeterd - can you double check if any last minute changes broke k8s startup?

streeterd commented:
@jakesmith I ran up the containerised build after having to exclude mongodb due to a compilation failure:

```
c/include -isystem /hpcc-dev/build/vcpkg_installed/x64-linux-dynamic/include/bsoncxx/v_noabi -isystem /hpcc-dev/build/vcpkg_installed/x64-linux-dynamic/include/mongocxx/v_noabi -frtti -fPIC -fmessage-length=0 -Werror=format -Wformat-security -Wformat-nonliteral -pthread -Wuninitialized -Werror=return-type -Werror=format-nonliteral -Wno-psabi -Wparentheses -g -D_DEBUG -DDEBUG -fPIC   -fsanitize=undefined -fno-sanitize=alignment -fsanitize-undefined-trap-on-error -MD -MT plugins/mongodb/CMakeFiles/mongodbembed.dir/mongodbembed.cpp.o -MF plugins/mongodb/CMakeFiles/mongodbembed.dir/mongodbembed.cpp.o.d -o plugins/mongodb/CMakeFiles/mongodbembed.dir/mongodbembed.cpp.o -c /hpcc-dev/HPCC-Platform/plugins/mongodb/mongodbembed.cpp
In file included from /hpcc-dev/build/vcpkg_installed/x64-linux-dynamic/include/bsoncxx/v_noabi/bsoncxx/stdx/optional.hpp:50,
                 from /hpcc-dev/build/vcpkg_installed/x64-linux-dynamic/include/bsoncxx/v_noabi/bsoncxx/document/element.hpp:27,
                 from /hpcc-dev/build/vcpkg_installed/x64-linux-dynamic/include/bsoncxx/v_noabi/bsoncxx/document/view.hpp:23,
                 from /hpcc-dev/build/vcpkg_installed/x64-linux-dynamic/include/mongocxx/v_noabi/mongocxx/client_session.hpp:28,
                 from /hpcc-dev/build/vcpkg_installed/x64-linux-dynamic/include/mongocxx/v_noabi/mongocxx/client.hpp:28,
                 from /hpcc-dev/HPCC-Platform/plugins/mongodb/mongodbembed.hpp:28,
                 from /hpcc-dev/HPCC-Platform/plugins/mongodb/mongodbembed.cpp:18:
/hpcc-dev/build/vcpkg_installed/x64-linux-dynamic/include/bsoncxx/v_noabi/bsoncxx/stdx/operators.hpp:88:34: error: 'bsoncxx::detail::strong_ordering::less' declared weak after being used
   88 |     static const strong_ordering less;
      |                                  ^~~~
/hpcc-dev/build/vcpkg_installed/x64-linux-dynamic/include/bsoncxx/v_noabi/bsoncxx/stdx/operators.hpp:89:34: error: 'bsoncxx::detail::strong_ordering::greater' declared weak after being used
   89 |     static const strong_ordering greater;
      |                                  ^~~~~~~

```
The pods run up ok
![image](https://github.com/user-attachments/assets/b10552c5-1422-4dac-b42c-4349a95e0940)

The log shows the debug housekeeping pod is ok:

```
dave@RISKAZ-06905526:~/HPCC-Platform/dockerfiles$ kl sasha-debugplane-housekeeping-6f7f868f64-wqr9b
running saserver --logging.postMortem=1000 --config=/etc/config/sasha-debugplane-housekeeping.yaml --daliServers=mydali:7070 --service=debugplane-housekeeping
Waiting for child process 7
00000000 USR PRO 2025-03-20 16:49:32.707     7     7 UNK     "JLog: queuing enabled: queueLen=512"
00000001 PRG INF 2025-03-20 16:49:32.707     7     7 UNK     "Build community_9.11.0-trunk0Debug[heads/HPCC-27051_Create_a_Sasha_service_to_clean_up_post-mortem_files-0-g7de389-dirty]"
00000002 USR PRO 2025-03-20 16:49:32.708     7     7 UNK     "MP TLS: off acceptThreadPoolSize: 100"
00000003 USR PRO 2025-03-20 16:49:34.915     7     7 UNK     "initializeStoragePlanes update"
00000004 PRG INF 2025-03-20 16:49:34.916     7     7 UNK     "Modify functions should be registered before the configuration is loaded"
00000005 USR PRO 2025-03-20 16:49:34.916     7     7 UNK     "SASERVER starting on sasha-debugplane-housekeeping-6f7f868f64-wqr9b:8877"
00000006 PRG INF 2025-03-20 16:49:34.916     7     7 UNK     "Creating sentinel file /tmp/sasha-debugplane-housekeeping.sentinel"
dave@RISKAZ-06905526:~/HPCC-Platform/dockerfiles$

```

I then completed a delete test of expired file and dir:

```
dave@RISKAZ-06905526:~$ kl sasha-debugplane-housekeeping-6f7f868f64-wqr9b
running saserver --logging.postMortem=1000 --config=/etc/config/sasha-debugplane-housekeeping.yaml --daliServers=mydali:7070 --service=debugplane-housekeeping
Waiting for child process 7
00000000 USR PRO 2025-03-20 16:49:32.707     7     7 UNK     "JLog: queuing enabled: queueLen=512"
00000001 PRG INF 2025-03-20 16:49:32.707     7     7 UNK     "Build community_9.11.0-trunk0Debug[heads/HPCC-27051_Create_a_Sasha_service_to_clean_up_post-mortem_files-0-g7de389-dirty]"
00000002 USR PRO 2025-03-20 16:49:32.708     7     7 UNK     "MP TLS: off acceptThreadPoolSize: 100"
00000003 USR PRO 2025-03-20 16:49:34.915     7     7 UNK     "initializeStoragePlanes update"
00000004 PRG INF 2025-03-20 16:49:34.916     7     7 UNK     "Modify functions should be registered before the configuration is loaded"
00000005 USR PRO 2025-03-20 16:49:34.916     7     7 UNK     "SASERVER starting on sasha-debugplane-housekeeping-6f7f868f64-wqr9b:8877"
00000006 PRG INF 2025-03-20 16:49:34.916     7     7 UNK     "Creating sentinel file /tmp/sasha-debugplane-housekeeping.sentinel"
00000007 USR PRO 2025-03-20 16:59:44.024     7    18 UNK     "DEBUGPLANEHOUSEKEEPING: Deleted dirs: 1, files: 1"
dave@RISKAZ-06905526:~$

```

Still running ok after that:
![image](https://github.com/user-attachments/assets/3f617551-5de1-41fe-bd87-4602a750b84e)

All looks ok.

jakesmith commented:
Not related to the k8s stall, but noticed some spurious white space in a couple of the yaml files in this PR:

![image](https://github.com/user-attachments/assets/443e5467-bcae-490d-8e2c-599a1be58634)

jakesmith reviewed (changes requested):
Pending on investigations into why k8s GH test failing.
Was failing because was [trying] to connect to Dali, but had no network policy to do so, and isn't needed. Making it conditional caused a seg. fault - to investigate.

jakesmith reviewed (changes requested):
@streeterd - passing this back to you for now, as BM regression suites are failing

jakesmith commented on dali/sasha/saserver.cpp:
```diff
@@ -238,6 +240,29 @@ class CSashaCmdThread : public CInterface, implements IPooledThread
 {
+    if (!isDaliClient)
+    {
+        bool stopped = false;
```
should really be an atomic.

jakesmith commented on dali/sasha/saserver.cpp:
```diff
@@ -238,6 +240,29 @@ class CSashaCmdThread : public CInterface, implements IPooledThread
+            virtual bool onAbort() override
+            {
+                PROGLOG("Aborted");
+                stopped = true;
```
This isn't going to affect the 'stopped' outside of the class, so it won't cause it to stop at the moment.
this.stopped needs to be a reference to the stopped passed in.

jakesmith commented on dali/sasha/saserver.cpp:
```diff
@@ -469,8 +498,11 @@ int main(int argc, const char* argv[])
+                            }
                         }
                     }
                 } *stopThread = new CStopThread;
```
unconditionally creating CStopThread here, but conditionally deleting it on line 521.
If not a Dali client, then doesn't need this thread.
It doesn't need to signal the stopSem either (doesn't need the stopSem at all, but harmless).

jakesmith reviewed (commented):
@streeterd - please see comments.

streeterd commented:
@jakesmith I tested the latest commit ba5a4ffcfc1a4f0635c8a18a25ce6e75f7f53aa4

In containerized:

```
running saserver --logging.postMortem=1000 --config=/etc/config/sasha-debugplane-housekeeping.yaml --daliServers=mydali:7070 --service=debugplane-housekeeping
Waiting for child process 7
00000000 USR PRO 2025-04-04 10:32:09.446     7     7 UNK     "JLog: queuing enabled: queueLen=512"
00000001 PRG INF 2025-04-04 10:32:09.446     7     7 UNK     "Build community_9.11.0-trunk0Debug[heads/HPCC-27051_Create_a_Sasha_service_to_clean_up_post-mortem_files-0-gba5a4f-dirty]"
00000002 USR PRO 2025-04-04 10:32:09.447     7     7 UNK     "initializeStoragePlanes update"
00000003 PRG INF 2025-04-04 10:32:09.447     7     7 UNK     "Modify functions should be registered before the configuration is loaded"
00000004 PRG INF 2025-04-04 10:32:09.447     7     7 UNK     "Creating sentinel file /tmp/sasha-debugplane-housekeeping.sentinel"
00000005 USR PRO 2025-04-04 10:39:50.037     7    13 UNK     "DEBUGPLANEHOUSEKEEPING: Deleted dirs: 1, files: 1"
EXIT via signal for 7
Sending SIGTERM to process (PID 7)
00000006 USR PRO 2025-04-04 10:40:23.751     7     7 UNK     "Stop signalled"
00000007 USR PRO 2025-04-04 10:40:23.751     7     7 UNK     "Stop requested"
00000008 USR PRO 2025-04-04 10:40:23.751     7     7 UNK     "SASERVER exiting"
00000009 USR PRO 2025-04-04 10:40:23.751     7     7 UNK     "Stopping 0"
Child process 7 exited with exit code 0
```

and

Bare metal:

```
dave@RISKAZ-06905526:~/HPCC-Platform$ bms
Starting mydafilesrv ...       [   OK    ]
Starting mydali ...            [   OK    ]
Starting mydfuserver ...       [   OK    ]
Starting myeclagent ...        [   OK    ]
Starting myeclccserver ...     [   OK    ]
Starting myeclscheduler ...    [   OK    ]
Starting myesp ...             [   OK    ]
Starting myroxie ...           [   OK    ]
Starting mysasha ...           [   OK    ]
Starting mythor ...            [   OK    ]
Starting mytoposerver ...      [   OK    ]
dave@RISKAZ-06905526:~/HPCC-Platform$ bme
Stopping mytoposerver...       [   OK    ]
Stopping mythor...             [   OK    ]
Stopping mysasha...            [   OK    ]
Stopping myroxie...            [   OK    ]
Stopping myesp...              [   OK    ]
Stopping myeclscheduler...     [   OK    ]
Stopping myeclccserver...      [   OK    ]
Stopping myeclagent...         [   OK    ]
Stopping mydfuserver...        [   OK    ]
Stopping mydali...             [   OK    ]
Service dafilesrv, mydafilesrv is still running.
To stop it, run "/etc/init.d/dafilesrv stop".
Stopping mydafilesrv...        [   OK    ]
dave@RISKAZ-06905526:~/hpccDebug/etc/init.d$

```

jakesmith commented on dali/sasha/sadbghk.cpp:
```diff
@@ -0,0 +1,179 @@
+        }
+        Owned<IPropertyTree> plane = getStoragePlane(planeName);
+        assertex(plane);
+        verifyex(plane->getProp("@prefix", debugDir));
```
trivial: reads odd that line 55 is an assertex and this is a verifyex. I'd change it to a single verifyex:
```
verifyex(plane && plane->getProp("@prefix", debugDir));
```

jakesmith commented on dali/sasha/sadbghk.cpp:
```diff
@@ -0,0 +1,179 @@
+                // Process expired directories for housekeeping
+                if (hasExpired(iFile))
+                {
+                    recursiveRemoveDirectory(&iFile);
```
recursiveRemoveDirectory should really return false if it fails.. but beyond scope of this PR.
But maybe add a comment on end of ++dirsDeleted; e.g. // should be really be conditional on success recursiveRemoveDirectory

streeterd replied:
Fixed in 341117ff45e2e48c46e8281ce781a1ad90ead3ba

jakesmith commented on dali/sasha/sadbghk.cpp:
```diff
@@ -0,0 +1,179 @@
+                if (hasExpired(iFile))
+                {
+                    iFile.remove();
+                    ++filesDeleted;
```
should be conditional on result of iFile.remove().
Could track failed delete count too.

streeterd replied:
Fixed in 341117ff45e2e48c46e8281ce781a1ad90ead3ba

jakesmith commented on dali/sasha/sasha.md:
```diff
@@ -0,0 +1,63 @@
+
+- This is a maintenance service that automatically manages files and directories on the debug plane using a configured retention period.
+- This is required to control the disc space used by post-mortem and other files on the debug plane.
+- This service is designed to execute on a schedule with no interaction other; a single log entry for a summary of any deleted files or directories will be generated.
```
>with no interaction other

with no other interaction?

streeterd replied:
Fixed in 341117ff45e2e48c46e8281ce781a1ad90ead3ba

jakesmith commented on dali/sasha/sasha.md:
```diff
@@ -0,0 +1,63 @@
+
+    |attribute|description|values|default|required|
+    |:--------|:----------|:-----|:------|:-------|
+    |disabled|Disable or enable this service.|false|false|optional|
```
what is 'values' column vs default?

streeterd replied:
'values' will be removed.

streeterd replied:
Fixed in 341117ff45e2e48c46e8281ce781a1ad90ead3ba

jakesmith reviewed (commented):
@streeterd - Final comments, nothing significant, could be merged now.

Please address and go ahead and squash.

jakesmith reviewed (approved):
@streeterd - looks good. Please squash.

streeterd commented:
@jakesmith  squashed

jakesmith reviewed (approved):
@streeterd - looks good.
